### ISSUE TITLE
Optimize attention mask creation for sample packing.

### ISSUE THREAD

**Issue by joecummings** [pytorch/torchtune#1130]
Based on some excellent investigative work by @ebsmothers, we can see that we are not being entirely efficient in our attention mask creation for sample packing. See this code snippet for a possible faster way to do this: https://gist.github.com/ebsmothers/707ee5a260b0e25b820a19b2e634e921
